src/time_series.c

```diff
@@ -0,0 +1,198 @@
+
+  // allocate (local) boundary flux storage
+  PetscCall(RDyAlloc(PetscReal, 3 * num_boundary_edges, &(rdy->time_series.boundary_fluxes.fluxes)));
+
+  PetscFunctionReturn(0);
+}
+
+// Initializes time series data storage.
+PetscErrorCode InitTimeSeries(RDy rdy) {
+  PetscFunctionBegin;
+  rdy->time_series = (RDyTimeSeriesData){0};
+  if (rdy->config.output.time_series.boundary_fluxes > 0) {
+    InitBoundaryFluxes(rdy);
+    PetscCall(TSMonitorSet(rdy->ts, WriteTimeSeries, rdy, NULL));
+  }
+  PetscFunctionReturn(0);
+}
+
+// Accumulates boundary fluxes on the given boundary from the given array of
+// fluxes on boundary edges.
+PetscErrorCode AccumulateBoundaryFluxes(RDy rdy, RDyBoundary *boundary, PetscInt num_edges, PetscReal fluxes[num_edges][3]) {
+  PetscFunctionBegin;
+  RDyTimeSeriesData *time_series = &rdy->time_series;
+  if (time_series->boundary_fluxes.fluxes) {
+    // figure out whether this boundary has an auto-generated BC
+    PetscInt  b              = boundary - rdy->boundaries;
+    PetscBool auto_generated = rdy->boundary_conditions[b].auto_generated;
+
+    // if not, accumulate fluxes locally
+    if (!auto_generated) {
+      PetscInt n = rdy->time_series.boundary_fluxes.offsets[b];
+      for (PetscInt e = 0; e < boundary->num_edges; ++e) {
+        PetscInt edge_id = boundary->edge_ids[e];
+        PetscInt cell_id = rdy->mesh.edges.cell_ids[2 * edge_id];
+        if (rdy->mesh.cells.is_local[cell_id]) {
+          time_series->boundary_fluxes.fluxes[n].water_mass += fluxes[e][0];
+          time_series->boundary_fluxes.fluxes[n].x_momentum += fluxes[e][1];
+          time_series->boundary_fluxes.fluxes[n].y_momentum += fluxes[e][2];
+          ++n;
+        }
+      }
+    }
+  }
+  PetscFunctionReturn(0);
+}
+
+static PetscErrorCode WriteBoundaryFluxes(RDy rdy, PetscInt step, PetscReal time) {
+  PetscFunctionBegin;
+
+  PetscInt num_local_edges  = rdy->time_series.boundary_fluxes.num_local_edges;
+  PetscInt num_global_edges = rdy->time_series.boundary_fluxes.num_global_edges;
+
+  // flux metadata (global edge ID, boundary ID, BC type)
+  PetscInt local_flux_md[3 * num_local_edges];
+
+  // flux data itself (mass, x-momentum, y-momentum, edge x normal, edge y normal)
+  PetscReal local_flux_data[5 * num_local_edges];
+
+  // gather local metadata/data
+  PetscInt n = 0;
+  for (PetscInt b = 0; b < rdy->num_boundaries; ++b) {
+    RDyCondition bc = rdy->boundary_conditions[b];
+    if (!bc.auto_generated) {  // exclude auto-generated boundary conditions
+      RDyBoundary *boundary = &rdy->boundaries[b];
+      for (PetscInt e = 0; e < boundary->num_edges; ++e) {
+        PetscInt edge_id = boundary->edge_ids[e];
+        PetscInt cell_id = rdy->mesh.edges.cell_ids[2 * edge_id];
+        if (rdy->mesh.cells.is_local[cell_id]) {
+          local_flux_md[3 * n]     = rdy->mesh.edges.global_ids[edge_id];
```

Comment:
I'm not certain if `rdy->mesh.edges.global_ids[edge_id];` has the correct ID when running in parallel. If it doesn't, the code should be fixed in another PR.

Reply:
I believe the objective here is to have the edge ID be in the natural order. Before distributing the mesh, we only have code that makes sure the natural ID of cells is known after the mesh is partitioned. I guess we need to add code before distributing the mesh, so the natural ID of vertices and edges is know after mesh partition.

Reply:
Opened #68